### DISCUSSION TITLE
Reasonable defaults ("convention over configuration") with built-in adapters

### DISCUSSION THREAD

**Discussion by q3769** [square/moshi#1645] [Ideas]
> Moshi refuses to serialize platform types (java.*, javax.*, and android.*) without a user-provided type adapter. This is intended to prevent you from accidentally locking yourself to a specific JDK or Android release.

First off, if that is true, why is `java.lang.String` or `java.util.List` supported as built-in types in the first place?

I guess the point to be made here is that "reasonable defaults" is a better way to go. i.e. the default built-in types need to be expanded, such that it should just "work" for most common cases:

- Java 8 Date Time API, in addition to `java.util.Date`
- `java.lang.Throwable`
- `java.lang.CharSequence`
- ...

The user should not have to "configure" in order to make the most basic things work. The current list of built-in adapters is too restricted to be reasonable.

**Comment by JakeWharton** [comment]
> First off, if that is true, why is java.lang.String or java.util.List supported as built-in types in the first place?

Because JSON has strings and lists.

**Reply by q3769** [reply]
Not exactly "lists". JSON has "arrays", which is sort of the point because that makes `java.util.List` is a reasonable default to be supported. So is Set, Map... as already supported, and the list should expand.

**Reply by JakeWharton** [reply]
Well, we disagree. It's one of the principles the library was founded on rather than continuing to maintain Gson which does many of the things you desire. I would suggest you try using Gson instead.

**Reply by q3769** [reply]
Well, if that is truly the "principle", then Set, Map, and Rfc3339DateJsonAdapter should be removed from the main-line distribution, and may be put into some different module. 

Thanks for the Gson suggestion. Currently using it, but unfortunately Gson/Jackson doesn't have "reasonable defaults" quite right for our use, either; so looking for other places and got to know Moshi....

**Reply by JakeWharton** [reply]
It's almost as if everyone has a different set of things they want from the library and it's our job to both pick ones which are reasonable for most while also offering customization to still do the things you want.

The only library that will work for 100% of your uses cases with 0 configuration is the one you write yourself.
